File: WordloomFrontend/next/src/modules/orbit/domain/media.ts
/**
 * Media Resource 类型定义和 API 客户端
 */

// ============================================================================
// 类型定义
// ============================================================================

export enum MediaEntityType {
  BOOKSHELF_COVER = 'bookshelf_cover',
  NOTE_COVER = 'note_cover',
  CHECKPOINT_MARKER = 'checkpoint_marker',
  IMAGE_BLOCK = 'image_block',
  OTHER_BLOCK = 'other_block',
}

/**
 * 统一的媒体资源类型
 */
export interface MediaResource {
  id: string;
  workspace_id: string;
  entity_type: MediaEntityType;
  entity_id: string;
  file_name: string;
  file_path: string;
  file_size?: number;
  mime_type?: string;
  width?: number;
  height?: number;
  display_order: number;
  is_thumbnail: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * 创建媒体资源的请求
 */
export interface CreateMediaResourceRequest {
  workspace_id: string;
  entity_type: MediaEntityType;
  entity_id: string;
  display_order?: number;
}

/**
 * 重新排列媒体资源的请求
 */
export interface ReorderMediaResourceRequest {
  entity_id: string;
  entity_type: MediaEntityType;
  order: string[]; // media_id 列表
}

// ============================================================================
// API 客户端函数
// ============================================================================

const API_BASE = '/api/orbit/media';

/**
 * 上传媒体资源
 */
export async function uploadMedia(
  file: File,
  workspace_id: string,
  entity_type: MediaEntityType,
  entity_id: string,
  display_order: number = 0
): Promise<MediaResource> {
  const formData = new FormData();
  formData.append('file', file);

  // 构建 URL 查询参数
  const params = new URLSearchParams({
    workspace_id,
    entity_type,
    entity_id,
    display_order: display_order.toString(),
  });

  console.log('[uploadMedia] Uploading to:', `${API_BASE}/upload?${params.toString()}`);

  const response = await fetch(`${API_BASE}/upload?${params.toString()}`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const errorText = await response.text();
    console.error('[uploadMedia] Error response:', errorText);
    throw new Error(`Failed to upload media: ${response.statusText} - ${errorText}`);
  }

  const data = await response.json();
  console.log('[uploadMedia] Success:', data);
  return data;
}

/**
 * 获取实体的所有媒体资源
 */
export async function getMediaByEntity(
  entity_type: MediaEntityType,
  entity_id: string
): Promise<MediaResource[]> {
  const params = new URLSearchParams({
    entity_type,
    entity_id,
  });

  const response = await fetch(`${API_BASE}?${params.toString()}`);

  if (!response.ok) {
    throw new Error(`Failed to fetch media: ${response.statusText}`);
  }

  return response.json();
}

/**
 * 获取单个媒体资源
 */
export async function getMedia(media_id: string): Promise<MediaResource> {
  const response = await fetch(`${API_BASE}/${media_id}`);

  if (!response.ok) {
    throw new Error(`Failed to fetch media: ${response.statusText}`);
  }

  return response.json();
}

/**
 * 删除媒体资源
 */
export async function deleteMedia(media_id: string): Promise<void> {
  const response = await fetch(`${API_BASE}/${media_id}`, {
    method: 'DELETE',
  });

  if (!response.ok) {
    throw new Error(`Failed to delete media: ${response.statusText}`);
  }
}

/**
 * 重新排列媒体资源
 */
export async function reorderMedia(
  payload: ReorderMediaResourceRequest
): Promise<void> {
  const response = await fetch(`${API_BASE}/reorder`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Failed to reorder media: ${response.statusText}`);
  }
}

File: WordloomBackend/api/app/schemas/orbit/media.py
"""
Media Resource 相关的 Pydantic 模式
"""
from pydantic import BaseModel
from datetime import datetime
from typing import Optional
from uuid import UUID
import enum


class MediaEntityType(str, enum.Enum):
    """媒体资源实体类型"""
    BOOKSHELF_COVER = "bookshelf_cover"
    NOTE_COVER = "note_cover"
    CHECKPOINT_MARKER = "checkpoint_marker"
    IMAGE_BLOCK = "image_block"
    OTHER_BLOCK = "other_block"


class MediaResourceCreate(BaseModel):
    """创建媒体资源的输入"""
    workspace_id: UUID
    entity_type: MediaEntityType
    entity_id: UUID
    file_name: str
    file_path: str
    file_size: Optional[int] = None
    mime_type: Optional[str] = None
    file_hash: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    display_order: Optional[int] = 0
    is_thumbnail: Optional[bool] = False


class MediaResourceUpdate(BaseModel):
    """更新媒体资源的输入"""
    display_order: Optional[int] = None
    is_thumbnail: Optional[bool] = None


class MediaResourceResponse(BaseModel):
    """媒体资源的输出"""
    id: UUID
    workspace_id: UUID
    entity_type: MediaEntityType
    entity_id: UUID
    file_name: str
    file_path: str
    file_size: Optional[int] = None
    mime_type: Optional[str] = None
    width: Optional[int] = None
    height: Optional[int] = None
    display_order: int
    is_thumbnail: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class MediaResourceDeleteRequest(BaseModel):
    """删除媒体资源的输入"""
    media_id: UUID


class MediaResourceReorderRequest(BaseModel):
    """重新排列媒体资源的输入"""
    entity_id: UUID
    entity_type: MediaEntityType
    order: list[UUID]  # 新的排列顺序 (media_id 列表)
